File: code/ts/src/graphql/schema/graphql_schema_sdl_generator.ts
import {
  GraphqlCatalogFieldVisibility,
  GraphqlCatalogRelationCardinality,
  type GraphqlCatalog,
  type GraphqlCatalogField,
  type GraphqlPublishedObject,
} from '../catalog/graphql_catalog_builder';

export class GraphqlSchemaSdlGenerator {
  generate(catalog: GraphqlCatalog): string {
    const objects = [...catalog.objects].sort((left, right) => left.id.localeCompare(right.id));
    const objectById = new Map(objects.map((object) => [object.id, object]));
    const blocks: string[] = [];

    const customScalars = this.collectCustomScalars(objects);
    if (customScalars.length > 0) {
      blocks.push(customScalars.map((scalar) => `scalar ${scalar}`).join('\n'));
    }

    blocks.push(this.buildQueryType(objects));

    for (const object of objects) {
      blocks.push(this.buildObjectType(object, objectById));
      blocks.push(this.buildListEnvelope(object));
      if (object.capabilities.item && object.graphql.itemField) {
        blocks.push(this.buildKeyInput(object));
      }
      blocks.push(this.buildFilterInput(object));
      blocks.push(this.buildOrderByInput(object));
      blocks.push(this.buildOrderFieldEnum(object));
    }

    for (const family of this.collectUsedFilterFamilies(objects)) {
      blocks.push(this.buildScalarFilterInput(family));
    }

    blocks.push(`enum SortDirection {
  ASC
  DESC
}`);
    blocks.push(`input OffsetPageInput {
  limit: Int
  offset: Int
}`);

    return blocks.join('\n\n');
  }

  private collectCustomScalars(objects: readonly GraphqlPublishedObject[]): string[] {
    const scalars = new Set<string>();

    for (const object of objects) {
      for (const field of object.fields) {
        if (this.isCustomScalar(field.type)) {
          scalars.add(field.type);
        }
      }
    }

    return [...scalars].sort((left, right) => left.localeCompare(right));
  }

  private buildQueryType(objects: readonly GraphqlPublishedObject[]): string {
    const lines = ['type Query {'];

    for (const object of objects) {
      if (object.capabilities.item && object.graphql.itemField) {
        lines.push(`  ${object.graphql.itemField}(key: ${object.graphql.typeName}KeyInput!): ${object.graphql.typeName}`);
      }
      if (object.capabilities.collection) {
        lines.push(`  ${object.graphql.collectionField}(`);
        lines.push(`    filter: ${object.graphql.typeName}FilterInput`);
        lines.push(`    orderBy: [${object.graphql.typeName}OrderByInput!]`);
        lines.push('    page: OffsetPageInput');
        lines.push(`  ): ${object.graphql.typeName}List!`);
      }
    }

    lines.push('}');
    return lines.join('\n');
  }

  private buildObjectType(
    object: GraphqlPublishedObject,
    objectById: ReadonlyMap<string, GraphqlPublishedObject>,
  ): string {
    const lines = [`type ${object.graphql.typeName} {`];

    for (const field of object.fields) {
      if (field.visibility !== GraphqlCatalogFieldVisibility.Public) {
        continue;
      }
      lines.push(`  ${field.publicName}: ${this.graphqlFieldType(field.type, field.nullable)}`);
    }

    for (const relation of object.relations) {
      const target = objectById.get(relation.target);
      if (!target) {
        continue;
      }
      const relationType =
        relation.cardinality === GraphqlCatalogRelationCardinality.Many
          ? `[${target.graphql.typeName}!]!`
          : target.graphql.typeName;
      lines.push(`  ${relation.name}: ${relationType}`);
    }

    lines.push('}');
    return lines.join('\n');
  }

  private buildListEnvelope(object: GraphqlPublishedObject): string {
    return `type ${object.graphql.typeName}List {
  items: [${object.graphql.typeName}!]!
  totalCount: Int!
}`;
  }

  private buildKeyInput(object: GraphqlPublishedObject): string {
    const fieldByColumn = new Map(object.fields.map((field) => [field.column, field]));
    const lines = [`input ${object.graphql.typeName}KeyInput {`];

    for (const column of object.identity.fields) {
      const field = fieldByColumn.get(column);
      if (!field) {
        continue;
      }
      lines.push(`  ${field.publicName}: ${field.type}!`);
    }

    lines.push('}');
    return lines.join('\n');
  }

  private buildFilterInput(object: GraphqlPublishedObject): string {
    const lines = [`input ${object.graphql.typeName}FilterInput {`];
    lines.push(`  and: [${object.graphql.typeName}FilterInput!]`);
    lines.push(`  or: [${object.graphql.typeName}FilterInput!]`);
    lines.push(`  not: ${object.graphql.typeName}FilterInput`);

    for (const field of object.fields) {
      if (
        field.visibility !== GraphqlCatalogFieldVisibility.Public ||
        !field.filterable ||
        field.type === 'Json'
      ) {
        continue;
      }
      lines.push(`  ${field.publicName}: ${field.type}FilterInput`);
    }

    lines.push('}');
    return lines.join('\n');
  }

  private buildOrderByInput(object: GraphqlPublishedObject): string {
    return `input ${object.graphql.typeName}OrderByInput {
  field: ${object.graphql.typeName}OrderField!
  direction: SortDirection!
}`;
  }

  private buildOrderFieldEnum(object: GraphqlPublishedObject): string {
    const lines = [`enum ${object.graphql.typeName}OrderField {`];

    for (const field of object.fields) {
      if (field.visibility !== GraphqlCatalogFieldVisibility.Public || !field.sortable) {
        continue;
      }
      lines.push(`  ${this.enumValueForFieldName(field.publicName)}`);
    }

    lines.push('}');
    return lines.join('\n');
  }

  private collectUsedFilterFamilies(objects: readonly GraphqlPublishedObject[]): string[] {
    const families = new Set<string>();

    for (const object of objects) {
      for (const field of object.fields) {
        if (
          field.visibility === GraphqlCatalogFieldVisibility.Public &&
          field.filterable &&
          field.type !== 'Json'
        ) {
          families.add(field.type);
        }
      }
    }

    return [...families].sort((left, right) => left.localeCompare(right));
  }

  private buildScalarFilterInput(scalar: string): string {
    switch (scalar) {
      case 'String':
        return `input StringFilterInput {
  eq: String
  ne: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
  isNull: Boolean
}`;
      case 'Boolean':
        return `input BooleanFilterInput {
  eq: Boolean
  ne: Boolean
  isNull: Boolean
}`;
      case 'Uuid':
        return `input UuidFilterInput {
  eq: Uuid
  ne: Uuid
  in: [Uuid!]
  isNull: Boolean
}`;
      case 'Int':
      case 'Long':
      case 'Float':
      case 'Decimal':
      case 'Date':
      case 'DateTime':
        return `input ${scalar}FilterInput {
  eq: ${scalar}
  ne: ${scalar}
  in: [${scalar}!]
  lt: ${scalar}
  lte: ${scalar}
  gt: ${scalar}
  gte: ${scalar}
  isNull: Boolean
}`;
      case 'Json':
        throw new Error('Json fields do not expose scalar filter operators in v1.');
      default:
        throw new Error(`Unsupported scalar family ${scalar} for filter input generation.`);
    }
  }

  private isCustomScalar(scalar: string): boolean {
    return ['Long', 'Decimal', 'Date', 'DateTime', 'Uuid', 'Json'].includes(scalar);
  }

  private graphqlFieldType(scalar: string, nullable: boolean): string {
    return nullable ? scalar : `${scalar}!`;
  }

  private enumValueForFieldName(publicName: string): string {
    const words: string[] = [];
    const segmentPattern = /[A-Za-z0-9]+/g;
    const wordPattern = /[A-Z]+(?:\d+)?(?=[A-Z][a-z]|$)|[A-Z]?[a-z]+\d*|\d+/g;
    const segments = publicName.match(segmentPattern) ?? [];

    for (const segment of segments) {
      const matches = segment.match(wordPattern) ?? [];
      for (const match of matches) {
        if (match) {
          words.push(match.toUpperCase());
        }
      }
    }

    return words.join('_');
  }
}

File: code/dart/modular_api/lib/src/graphql/schema/graphql_schema_sdl_generator.dart
import 'package:modular_api/src/graphql/catalog/graphql_catalog_builder.dart';

final class GraphqlSchemaSdlGenerator {
  const GraphqlSchemaSdlGenerator();

  String generate(GraphqlCatalog catalog) {
    final objects = catalog.objects.toList(growable: false)
      ..sort((left, right) => left.id.compareTo(right.id));
    final objectById = <String, GraphqlPublishedObject>{
      for (final object in objects) object.id: object,
    };

    final blocks = <String>[];
    final customScalars = _collectCustomScalars(objects);
    if (customScalars.isNotEmpty) {
      blocks.add(customScalars.map((scalar) => 'scalar $scalar').join('\n'));
    }

    blocks.add(_buildQueryType(objects));

    for (final object in objects) {
      blocks.add(_buildObjectType(object, objectById));
      blocks.add(_buildListEnvelope(object));
      if (object.capabilities.item && object.graphql.itemField != null) {
        blocks.add(_buildKeyInput(object));
      }
      blocks.add(_buildFilterInput(object));
      blocks.add(_buildOrderByInput(object));
      blocks.add(_buildOrderFieldEnum(object));
    }

    final usedFilterFamilies = _collectUsedFilterFamilies(objects);
    for (final family in usedFilterFamilies) {
      blocks.add(_buildScalarFilterInput(family));
    }

    blocks.add('''
enum SortDirection {
  ASC
  DESC
}''');
    blocks.add('''
input OffsetPageInput {
  limit: Int
  offset: Int
}''');

    return blocks.join('\n\n');
  }

  List<String> _collectCustomScalars(List<GraphqlPublishedObject> objects) {
    final usedScalars = <String>{};
    for (final object in objects) {
      for (final field in object.fields) {
        if (_isCustomScalar(field.type)) {
          usedScalars.add(field.type);
        }
      }
    }

    final scalars = usedScalars.toList(growable: false)..sort();
    return scalars;
  }

  String _buildQueryType(List<GraphqlPublishedObject> objects) {
    final lines = <String>['type Query {'];

    for (final object in objects) {
      if (object.capabilities.item && object.graphql.itemField != null) {
        lines.add(
          '  ${object.graphql.itemField}(key: ${object.graphql.typeName}KeyInput!): ${object.graphql.typeName}',
        );
      }
      if (object.capabilities.collection) {
        lines.add('  ${object.graphql.collectionField}(');
        lines.add('    filter: ${object.graphql.typeName}FilterInput');
        lines.add('    orderBy: [${object.graphql.typeName}OrderByInput!]');
        lines.add('    page: OffsetPageInput');
        lines.add('  ): ${object.graphql.typeName}List!');
      }
    }

    lines.add('}');
    return lines.join('\n');
  }

  String _buildObjectType(
    GraphqlPublishedObject object,
    Map<String, GraphqlPublishedObject> objectById,
  ) {
    final lines = <String>['type ${object.graphql.typeName} {'];
    for (final field in object.fields.where(
      (field) => field.visibility == GraphqlCatalogFieldVisibility.public,
    )) {
      lines.add('  ${field.publicName}: ${_graphqlFieldType(field.type, field.nullable)}');
    }
    for (final relation in object.relations) {
      final target = objectById[relation.target];
      if (target == null) {
        continue;
      }
      final relationType = relation.cardinality == GraphqlCatalogRelationCardinality.many
          ? '[${target.graphql.typeName}!]!'
          : target.graphql.typeName;
      lines.add('  ${relation.name}: $relationType');
    }
    lines.add('}');
    return lines.join('\n');
  }

  String _buildListEnvelope(GraphqlPublishedObject object) {
    return '''
type ${object.graphql.typeName}List {
  items: [${object.graphql.typeName}!]!
  totalCount: Int!
}''';
  }

  String _buildKeyInput(GraphqlPublishedObject object) {
    final fieldByColumn = <String, GraphqlCatalogField>{
      for (final field in object.fields) field.column: field,
    };
    final lines = <String>['input ${object.graphql.typeName}KeyInput {'];
    for (final column in object.identity.fields) {
      final field = fieldByColumn[column];
      if (field == null) {
        continue;
      }
      lines.add('  ${field.publicName}: ${field.type}!');
    }
    lines.add('}');
    return lines.join('\n');
  }

  String _buildFilterInput(GraphqlPublishedObject object) {
    final lines = <String>['input ${object.graphql.typeName}FilterInput {'];
    lines.add('  and: [${object.graphql.typeName}FilterInput!]');
    lines.add('  or: [${object.graphql.typeName}FilterInput!]');
    lines.add('  not: ${object.graphql.typeName}FilterInput');
    for (final field in object.fields.where(
      (field) =>
          field.visibility == GraphqlCatalogFieldVisibility.public &&
          field.filterable &&
          field.type != 'Json',
    )) {
      lines.add('  ${field.publicName}: ${field.type}FilterInput');
    }
    lines.add('}');
    return lines.join('\n');
  }

  String _buildOrderByInput(GraphqlPublishedObject object) {
    return '''
input ${object.graphql.typeName}OrderByInput {
  field: ${object.graphql.typeName}OrderField!
  direction: SortDirection!
}''';
  }

  String _buildOrderFieldEnum(GraphqlPublishedObject object) {
    final lines = <String>['enum ${object.graphql.typeName}OrderField {'];
    for (final field in object.fields.where(
      (field) =>
          field.visibility == GraphqlCatalogFieldVisibility.public && field.sortable,
    )) {
      lines.add('  ${_enumValueForFieldName(field.publicName)}');
    }
    lines.add('}');
    return lines.join('\n');
  }

  List<String> _collectUsedFilterFamilies(List<GraphqlPublishedObject> objects) {
    final families = <String>{};
    for (final object in objects) {
      for (final field in object.fields) {
        if (field.visibility == GraphqlCatalogFieldVisibility.public &&
            field.filterable &&
            field.type != 'Json') {
          families.add(field.type);
        }
      }
    }
    final sorted = families.toList(growable: false)..sort();
    return sorted;
  }

  String _buildScalarFilterInput(String scalar) {
    switch (scalar) {
      case 'String':
        return '''
input StringFilterInput {
  eq: String
  ne: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
  isNull: Boolean
}''';
      case 'Boolean':
        return '''
input BooleanFilterInput {
  eq: Boolean
  ne: Boolean
  isNull: Boolean
}''';
      case 'Uuid':
        return '''
input UuidFilterInput {
  eq: Uuid
  ne: Uuid
  in: [Uuid!]
  isNull: Boolean
}''';
      case 'Int':
      case 'Long':
      case 'Float':
      case 'Decimal':
      case 'Date':
      case 'DateTime':
        return '''
input ${scalar}FilterInput {
  eq: $scalar
  ne: $scalar
  in: [$scalar!]
  lt: $scalar
  lte: $scalar
  gt: $scalar
  gte: $scalar
  isNull: Boolean
}''';
      case 'Json':
        throw StateError('Json fields do not expose scalar filter operators in v1.');
    }
    throw StateError('Unsupported scalar family $scalar for filter input generation.');
  }

  bool _isCustomScalar(String scalar) {
    return switch (scalar) {
      'Long' || 'Decimal' || 'Date' || 'DateTime' || 'Uuid' || 'Json' => true,
      _ => false,
    };
  }

  String _graphqlFieldType(String scalar, bool nullable) {
    return nullable ? scalar : '$scalar!';
  }

  String _enumValueForFieldName(String publicName) {
    final words = <String>[];
    final segmentPattern = RegExp(r'[A-Za-z0-9]+');
    final wordPattern = RegExp(
      r'[A-Z]+(?:\d+)?(?=[A-Z][a-z]|$)|[A-Z]?[a-z]+\d*|\d+',
    );

    for (final segmentMatch in segmentPattern.allMatches(publicName)) {
      final segment = segmentMatch.group(0)!;
      for (final wordMatch in wordPattern.allMatches(segment)) {
        final word = wordMatch.group(0)!;
        if (word.isNotEmpty) {
          words.add(word.toUpperCase());
        }
      }
    }

    return words.join('_');
  }
}